worker/undertaker/manifold.go

```diff
@@ -5,56 +5,76 @@ package undertaker
 
 import (
 	"github.com/juju/errors"
-	worker "gopkg.in/juju/worker.v1"
+	"gopkg.in/juju/worker.v1"
 
 	"github.com/juju/juju/api/base"
+	"github.com/juju/juju/caas"
 	"github.com/juju/juju/environs"
 	"github.com/juju/juju/worker/dependency"
 )
 
 // ManifoldConfig holds the names of the resources used by, and the
 // additional dependencies of, an undertaker worker.
 type ManifoldConfig struct {
-	APICallerName string
-	EnvironName   string
+	APICallerName      string
+	CloudDestroyerName string
 
 	NewFacade func(base.APICaller) (Facade, error)
 	NewWorker func(Config) (worker.Worker, error)
 }
 
-func (config ManifoldConfig) start(context dependency.Context) (worker.Worker, error) {
+func (config ManifoldConfig) startWorker(context dependency.Context, destroyer cloudDestroyer) (worker.Worker, error) {
 
 	var apiCaller base.APICaller
 	if err := context.Get(config.APICallerName, &apiCaller); err != nil {
 		return nil, errors.Trace(err)
 	}
-	var environ environs.Environ
-	if err := context.Get(config.EnvironName, &environ); err != nil {
-		return nil, errors.Trace(err)
-	}
-
 	facade, err := config.NewFacade(apiCaller)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 	worker, err := config.NewWorker(Config{
-		Facade:  facade,
-		Environ: environ,
+		Facade:    facade,
+		Destroyer: destroyer,
 	})
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 	return worker, nil
 }
 
-// Manifold returns a dependency.Manifold that runs a worker responsible
-// for shepherding a Dying model into Dead and ultimate removal.
-func Manifold(config ManifoldConfig) dependency.Manifold {
+// IAASManifold returns a dependency.Manifold that runs a worker responsible
+// for shepherding a Dying IAAS model into Dead and ultimate removal.
+func IAASManifold(config ManifoldConfig) dependency.Manifold {
+	return dependency.Manifold{
+		Inputs: []string{
+			config.APICallerName,
+			config.CloudDestroyerName,
+		},
+		Start: func(context dependency.Context) (worker.Worker, error) {
+			var environ environs.Environ
```

Comment:
Why not move the `CloudDestroyer`  definition somewhere public and pass a `*CloudDestroyer` into `context.Get`? Then you'd only need one manifold function, rather than two that are the same other than the type of the cloud destroyer - it's a bit weird that they have to know the type of the environ tracker they're talking to. You'd need to change the environ tracker  and caas broker workers to handle that output type, but I think it would be clearer.

Reply:
Good idea, thank you.